모임 생성 기능 구현

hyeonnkang opened the pull request:
## 작업 사항
close #17
- 모임 생성 기능 api 까지 구현 완료했습니다.
- 현재 MeetService의 createMeet 메소드 내부에서 동네 목록을 가져올 때 orElseThrow가 잘 되지 않아 현재 Optional 을 get으로 가져오는 방법을 사용했습니다. 피드백 주시면 수정할 수 있도록 하겠습니다!

## 참고 자료

- 참고자료 (링크 등)

## 체크리스트

- [x]  PR의 제목은 팀 내 규칙을 준수하여 알맞게 작성하였는가?
- [x]  Merge 하는 브랜치가 올바른가? (`main` branch에 실수로 PR 생성 금지)
- [x]  팀의 코딩 컨벤션을 준수하는가?
- [ ]  PR과 관련없는 변경사항이 들어가지는 않았는가?
- [x]  내 코드에 대한 자기 검토가 되었는가?
- [x]  Reviewers, Assignees, Lables, Project, Milestone은 적절하게 선택하였는가?
- [x]  관련한 issue를 닫아야 하는지 점검해보고 적용했는가?

Commit:
feat: #17 MeetRepository, MemberRepository 구현

MeetRepository 와 MemberRepository 를 JpaRepository를 이용해 생성했다. 추후 추가적인 메소드를 생성할 예정이다.

Commit:
feat: #17 MeetRepository 전체 모임 조회 기능 테스트

MeetRepository에서 전체 모임을 조회하는 기능을 테스트했다. 
모임 목록들의 갯수와 내용들이 제대로 들어갔는지 확인하여 조회 여부를 판단했다.
모임의 필드 값이 제대로 입력됬는지 확인하기 위해 Meet.java 에 toString 함수를 추가적으로 구현했다.

Commit:
feat: #17 MeetDto, MeetLeaderDto 생성

추후 제작될 MeetService와 MeetController 간의 사용될 Dto를 생성했다. 
현재 각각 두 Dto에 from 메소드와 toEntity 메소드를 생성하였으며, of 메소드는 추후 제작할 예정이다.

Commit:
feat: #17 'Meet'  전체조회 및 생성 기능 구현 및 테스트

모임을 조회 시 컨트롤러에 MeetDto와 MeetLeaderDto를 이용해 반환하도록 구현했고,
모임 생성 시에는 Meet 엔티티를 입력받아 저장하는 방식으로 구현했다.

Commit:
style: #17 불필요한 import문 제거

hyeonnkang committed:
feat: #17 'Meet' Id로 1개 조회 Repository Test

Repository Layer에서 Meet Id가 주어졌을 때 조회를 잘 하는지 테스트했다.

hyeonnkang committed:
feat: #17 'Meet' Id로 1개 조회 Service 기능 구현 및 Test

Service Layer에서 Meet을 1개 조회하는 기능을 구현 및 테스트 했다.

hyeonnkang committed:
feat: #17 'Meet' 제목으로 모임 리스트 조회 구현 및 테스트

Repository 및 Service Layer에서 제목으로 모임 리스트 조회 기능 추가 및 테스트했다. 
테스트는 Service Layer에서 테스트했다.

hyeonnkang committed:
refactor: #17 MeetServiceTest 리팩토링

함수 내에서 데이터 셋업하는 코드가 중복되어 이 부분을 따로 함수로 빼주었다.

hyeonnkang committed:
feat: #17 'MeetAddress' 엔티티 생성

hyeonnkang committed:
feat: #17 MeetDto 및 MeetLeaderDto of 메소드 구현

각 Dto의 모든 필드를 파라미터로 입력받아 Dto 인스턴스를 반환해주는 of 메소드를 구현했다. 추후에 일부 파라미터를 입력받는 of 메소드는 필요에 따라 추가적으로 구현 예정이다.

hyeonnkang committed:
feat: #17 'Meet' 카테고리로 조회 구현 및 테스트

Service Layer에서 카테고리로 모임글 리스트를 조회하는 기능을 구현 및 테스트 완료했다.

hyeonnkang committed:
feat: #17 'Meet' 동네 Address로 모임 리스트 조회 구현 및 테스트

Service Layer에서 동네로 모임 리스트 조회 기능을 구현했다.
MeetDto 에 설정된 address 목록을 넣는 것은 추후 구현 예정이다.

hyeonnkang committed:
feat: #17 MeetDto 에 동네 목록 넣기 및 테스트

MeetDto에 MeetAddressDto 목록 필드를 생성해서 모임의 동네 목록 리스트가 저장되고, 조회될 수 있도록 기능을 추가했다.

hyeonnkang committed:
feat: #17 MeetService 및 Reposiory 조회 삭제

MeetService 에서 조회 메소드를 하나만 통일하기 위해 현재 만들어둔 조회 메소드를 다 삭제했다.

hyeonnkang committed:
feat: #17 MeetRepositoryTest 삭제

MeetRepository에서 쿼리를 직접 짠 메소드가 없으므로 MeetRepositoryTest 를 삭제했다. 추후에 직접 쿼리를 짤 경우 그때 테스트를 진행할 예정이다.

hyeonnkang committed:
Merge branch 'develop' into feature/#17-meet-create

hyeonnkang committed:
feat: #17 toString 메소드 삭제 및 MeetServiceTest 메소드 삭제

Meet 엔티티 및 관련 DTO 에 있던 불필요한 toString 메소드를 삭제했다.
이전에 MeetService에서 조회 로직을 모두 삭제했기 때문에 MeetServiceTest 안에 있는 테스트 메소드들을 모두 삭제했다. 나중에 사용하기 위해 기본 데이터 셋업 함수는 남겨놓았다.

hyeonnkang committed:
feat: #17 패키지 이름 변경 및 Meet 관련 DTO 수정

DTO에서 반복적으로 구현되는 메소드에 대해 주석을 삭제했다.

MeetDTO에서 불필요하게 MeetAddressRepository Autowired 하는 부분이 있어서 이 부분을 삭제했다.

DTO에 Builder를 없애고 AllArgsConstructor를 private 접근으로 변경했다.

hyeonnkang committed:
refactor: #17 MeetDto 필드 위치 변경

hyeonnkang committed:
feat: #17 모임 생성 기능 구현

모임 생성 기능을 서비스 및 API 구현 완료했다. 
MeetService의 createMeet 메소드 내부에서 동네목록 리스트 설정을 할 때 
map 사용이 익숙치 않아 Optional 을 get으로 받아오는 방법을 사용한 상태인데, 추후 더 나은 방법으로 고칠 예정이다.

Wo-ogie commented on src/main/java/com/umc/withme/controller/MeetController.java:
```diff
@@ -0,0 +1,55 @@
+    private final MemberService memberService;
+
+    @PostMapping("/meet")
+    public ResponseEntity<DataResponse> createMeet(@Validated @RequestBody MeetCreateRequest meetCreateRequest){
```
`@RequestBody`에 대한 검증이 필요하신거면 `@Valid`만으로도 충분할 것 같습니다. 그리고 이 경우 class level에 달려있는 `@Validated`는 필요 없을 것 같아요.

Wo-ogie commented on src/main/java/com/umc/withme/dto/meet/MeetAddressDto.java:
```diff
@@ -0,0 +1,25 @@
+    public static MeetAddressDto of(String sido, String sgg){
+        return new MeetAddressDto(sido, sgg);
+    }
+}
```
1. DTO의 내용이 `MeetAddressDto` 보다는 `AddressDto`가 적합할 것 같습니다. `MeetAddress`에 대한 내용을 담고 있는 dto가 아닌, `Address`에 대한 내용을 담고 있는 dto니까요. 쓰고 계신 곳이 `MeetAddress`와 관련한 곳이라서 이렇게 이름 지은 것 같은데, `AddressDto`가 딱 이 모양일 것 같네요? 사실 저도 지금 이거랑 똑같은 `AddressDto`를 만들어서 쓰고 있는데, `MeetAddressDto`보다는 `AddressDto`가 적합할 것 같습니다.
2. `Address` entity는 저희가 더 생성할 일이 없습니다. 따라서 `Address` entity를 생성하는 코드는 필요하지도 않고, 구현하면 안됩니다. 이러한 이유 때문에 `toEntity()`는 삭제 해주시는 것이 좋을 것 같습니다.
3. [이전 코드리뷰](https://github.com/UMC-WithMe/WithMe_Back/pull/34#discussion_r1080835952)에도 비슷한 내용이 있는데, 객체를 생성하는 `of()` method가 존재하는 상황이니 `from()`에서는 생성자 대신 `of()` 메서드를 활용해서 객체를 생성해보는 것은 어떨까요?

Wo-ogie commented on src/main/java/com/umc/withme/domain/Meet.java:
```diff
@@ -67,6 +66,8 @@ private Meet(Member leader, MeetCategory category, String title, Integer minPeop
         this.link = link;
         this.content = content;
+        this.startDate = startDate;
+        this.endDate = endDate;
```
모임 게시글 등록 시 시작일과 종료일은 받지 않는 걸로 알고 있는데... 아닐까요?
그래서 `startDate`, `endDate`를 nullable로 했고, 시작 일시와 종료 일시는 모임 인원 모집이 완료되면 그 때 입력받는 것으로 알고 있습니다.

Wo-ogie commented on src/main/java/com/umc/withme/dto/meet/MeetCreateRequest.java:
```diff
@@ -0,0 +1,58 @@
+    private MeetCategory meetCategory;
+    @Nullable
+    private RecruitStatus recruitStatus;
+    private MeetStatus meetStatus;
```
`recruitStatus`는 entity 생성 시 `PROGRESS` 값으로 설정되도록 생성자가 구현되어 있고, 화면 설계서에도 모집 상태를 입력받고 있지는 않은 것 같습니다.
마찬가지로 `meetStatus`는 인원 모집 후 정해지는 값이라 처음에는 null이 저장될 것 입니다. 모임 생성 시 요청 데이터로 받을 필요가 없지 않을까요?

Wo-ogie commented on src/main/java/com/umc/withme/dto/meet/MeetDto.java:
```diff
@@ -0,0 +1,126 @@
+public class MeetDto {
+
+    private Long meetId;
+    private MeetLeaderDto leader;  // 이 모임의 리더
```
[디코 코드리뷰]
`MemberDto`로 변경

Wo-ogie commented on src/main/java/com/umc/withme/dto/meet/MeetDto.java:
```diff
@@ -0,0 +1,126 @@
+        this.maxPeople = maxPeople;
+        this.startDate = startDate;
+        this.endDate = endDate;
+    }
```
[디코 코드리뷰]
삭제해주세요

Wo-ogie commented on src/test/java/com/umc/withme/service/MeetServiceTest.java:
```diff
@@ -0,0 +1,79 @@
+@DisplayName("Meet - Service Layer Test")
+@SpringBootTest
+@Transactional
+class MeetServiceTest {
```
[디코 코드리뷰]
`@Disabled("추후에 작성할 테스트임")`

Wo-ogie commented on src/main/java/com/umc/withme/dto/meet/MeetDto.java:
```diff
@@ -0,0 +1,126 @@
+                meetCategory, recruitStatus, meetStatus,
+                title, link, content, minPeople, maxPeople,
+                startDate, endDate);
+        meetDto.setAddresses(addresses);
```
`this.addresses = addresses`

hyeonnkang committed:
feat: #17 코드리뷰 - 불필요한 메소드 삭제 및 Response 데이터 변경


우기가 해주신 피드백 모두 반영해서 수정했습니다.
1. @Validated -> @Valid로 변경
2. MeetAddressDto -> AddressDto로 변경
3. MeetDto의 of 메소드 변경
4. 모임글 생성 시 시작일과 마감일 받지 않도록 변경
5. MeetLeaderDto 삭제 및 MemberDto 사용으로 대체
6. MeetDto의 불필요한 생성자 삭제
7. 사용하지 않는 테스트 클래스에 @Disabled 추가

hyeonnkang committed:
feat: #17 코드리뷰 - 2번째 변경

hyeonnkang committed:
feat: #17 코드리뷰 - 3번째 변경

Wo-ogie committed:
feat: #17 코드리뷰 수정

원래 commit을 각각 분리해서 좀 더 자세하게 해야 하지만.... 한 번만 봐주시면 감사하겠습니다.

Wo-ogie committed:
feat: #17 Meet Entity를 기반으로 MeetDto 생성 로직 변경

기존에는 `new ArrayList()`로 그냥 임의 처리만 해주었다. 이 로직은 실제 사용되기에는 문제를 야기할 소지가 많으므로 이를 보완하였다.

`from()` method의 parameter로 모임에 등록된 address list를 받는다. 전달받은 address list를 활용하여 address dto list를 생성하도록 로직을 구현하였다.

Wo-ogie reviewed (approved):
제가 보기에는 크게 문제 없어보이는데.... postman으로 테스트 꼼꼼히  해주시고, 문제 없으면 merge 하셔도 될 것 같습니다.
수고하셨어요! 👍
